Added a test case

Bengt-M opened the pull request:
Adding on to @pqueiros excellent start.
And did some cleaning.
Had to pull some definitions to the header file for visibility.
(More cleaning might be good.)

Look at the test result. I think it demonstrates that the theory in #25 (K = 1/tan(A)) is right. Current tuning code can be simplified.

Bengt-M committed:
Added a test case

And did some cleaning.
Had to pull some definitions to the header file.
More cleaning might be good.

pqueiros commented on src/test/unit/mixer_tricopter_unittest.cc:
```diff
@@ -21,55 +21,162 @@
+}
+
+void beeperConfirmationBeeps(uint8_t beepCount) {
+    (void) beepCount;
```
Thank you for fixing this ones, To be honest learned something new today :).

Bengt-M replied:
An alternative would be to remove the variable name. I think this would work too:
uint16_t disableFlightMode(flightModeFlags_e /\* mask */) { 
Its a matter of taste.

lkaino replied:
There's actually a macro to define these, UNUSED(x)

pqueiros commented on src/main/flight/mixer_tricopter.c:
```diff
@@ -494,7 +418,7 @@ static void triTailTuneStep(servoParam_t *pServoConf, int16_t *pServoVal)
 }
 
-static void tailTuneModeThrustTorque(struct thrustTorque_t *pTT, const bool isThrottleHigh)
+void tailTuneModeThrustTorque(thrustTorque_t *pTT, const bool isThrottleHigh)
```
I would suggest to use STATIC_UNIT_TESTED for internal methods that require unit testing. We should hide as much as possible implementation details from the outside world.

lkaino replied:
I agree, let's keep the static functions and variables static.

Bengt-M replied:
Yes agree completely. I rememberd I saw something like that but couldn't find it.

pqueiros replied:
fyi.
https://github.com/cleanflight/cleanflight/blob/e133331cf34f94109774181c2d6f3ae74f4fba05/src/main/build_config.h

pqueiros commented on src/main/flight/mixer_tricopter.h:
```diff
@@ -73,4 +73,82 @@ int16_t triGetMotorCorrection(uint8_t motorIndex);
  */
 _Bool triEnableServoUnarmed(void);
 
+#ifdef USE_SERVOS
```
These are all details of the tricopter mixer implementation right? Which now need to be visible in the unit tests.
To keep a clear view on what is part of our public API and what is implementation details I would like to propose to move internals to a separate header file, eg mixer_tricopter_internals.h.

Bengt-M replied:
I was thinking along that path too, for a short while. I wanted to get a test running quickly so I dropped it.
Tried to follow the example from other units.
There is nothing dangerous about exposing type definitions, but variables and functions should be on a need-to-know basis (but without duplicating the same code lines).
Have to look again. Actually, this is what I was thinking of when I wrote "More cleaning might be good" in the PR comment.

pqueiros replied:
I understand, I think it is ok to do the cleaning later on as well. This should not be a blocker regarding the pull request.

lkaino replied:
I have used two different ways in the past with unit testing frameworks that don't support accessing statics or including the SUT .c-file in the test file:
1. split them to a separate *_internal.h header file
2. declare them in the public header file but enclose as follows:

```
#ifdef MIXER_TRICOPTER_INTERNALS
declare stuff
#endif

```

Then when including the header file in mixer_tricopter.c or in the test file you would have to

```
#define MIXER_TRICOPTER_INTERNALS
#include "mixer_tricopter.h"
```

This guards against unintentional external usage.

Bengt-M replied:
Wow you are quick.
Thanks for the input. I think your second pattern is really nice. No extra files.
(Now time to do my day job..)

lkaino commented on src/test/unit/mixer_tricopter_unittest.cc:
```diff
@@ -21,55 +21,162 @@
+TEST_F(BasicTest, TestThrustFactorCalculation139) {
+    // given
+    // expecting tri_tail_motor_thrustfactor = 1 / tan(angle)
+    tailTune.tt.servoAvgAngle.sum = 12345;
```
Could you comment on this first test case how did you come with the 12345, maybe calculate it here.

Bengt-M replied:
Yeh, the number looks funny.
I want the result to be 139 (i.e. factor is 13.9)
As per #25 I can calculate the corresponding angle by
atan(1/13.9) = 4.115 deg
Change it to decidegrees and multiply with the servoAvgAngle.numOf
Together that means multiply by 3000
So servoAvgAngle.sum should be 12344.73 but I round it to nearest integer.

I didn't try to analyse if rounding up or down is more correct (as that is part of the implementation and not the requirement) so I but a tolerance of 1 (which is pretty tight requirement anyway).

lkaino replied:
I understand, I don't want to be a nitpicker here, I'm so grateful that you wrote unit tests. Something that I should have done.

Just thinking that we could comment the formula in this first test case so the next one who reads through understands where did the magic number come from.

pqueiros commented:
@Bengt-M It is great to see someone started actually using the unit tests to support implementation :).
Thanks for fixing some of the stuff I pushed :).
I have done some suggestions that you might want to take a look.

lkaino commented:
Thanks guys, could you @Bengt-M check the comments?

Bengt-M committed:
Updated after review comments

Thanks to @lkaino and @pqueiros.
Also did some naming improvements and some more cleaning up (K&R).

Bengt-M commented:
New commit made

lkaino commented on src/test/unit/mixer_tricopter_unittest.cc:
```diff
@@ -21,55 +21,174 @@
-    ASSERT_TRUE(false);
+TEST_F(ThrustFactorCalculationTest, 139) {
+    // given
+    tailTune.tt.servoAvgAngle.sum = 12345;
```
The real world angles are > 90 degrees when motor is spinning CCW (sum > 270000). Why did you choose such a small numbers for the tests?

Bengt-M replied:
In my experience the angle is around 4 degrees, but it depends on where 0
is. Maybe I got it wrong.
Did the test pass or fail for you?

2016-05-03 19:31 GMT+02:00 lkaino notifications@github.com:

> In src/test/unit/mixer_tricopter_unittest.cc
> https://github.com/lkaino/Triflight/pull/27#discussion_r61921727:
> 
> > -TEST(ATest, test) {
> > -    ASSERT_TRUE(false);
> >   +TEST_F(ThrustFactorCalculationTest, 139) {
> > -    // given
> > -    tailTune.tt.servoAvgAngle.sum = 12345;
> 
> The real world angles are > 90 degrees when motor is spinning CCW (sum >
> 270000). Why did you choose such a small numbers for the tests?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/lkaino/Triflight/pull/27/files/50a3f8cf04fa74f453d06a105602c521a83be739#r61921727

lkaino replied:
It is around 4 degrees from middle. In this case middle is 90 degrees.

Bengt-M replied:
Very interesting. I added 270000 (90 deg) to all angles and committed again.
All tests pass again. Your functions must somehow be periodic.
(Probably means that the alternative implementation also has to compensate 90 deg, but I haven't tried that yet.)

Bengt-M replied:
@lkaino; First this was hard to believe but very good you noticed. I think you are (almost) right about where the 0 degrees is but the function produce the same data anyway. Look here: [Book1.pdf](https://github.com/lkaino/Triflight/files/247650/Book1.pdf)

This graph shows your function in blue. Data taken from the code itself using a temp test case iterating a lot of printf.
It seems it will produce the same result in the region 1.5-20 as in 91.5-110 degrees.
The black dashed lines are what the simpler implementation will do. Current one to the left and after (still pending) correction on the right.

(About the "almost" above: I don't think the position at 90 deg is exactly vertical for the servo. It should be where the thrust force is vertical. Almost the same but not exactly.)

lkaino replied:
Thanks for going through that implementation, it seems that your version does even better near 90 degrees!

Bengt-M committed:
Fix coordinate bug
